Adding various Canvas functions

star-nox opened the pull request:
added function for getting canvas user emails

Here's our development course: https://canvas.illinois.edu/courses/37348

KastanDay reviewed (approved):
I didn’t test it, but the code structure & quality looks very nice.

KastanDay commented:
How do we know we were successful?
1. All documents on Canvas are uploaded to course.
  a. On cron job or on-event listener (often called a webhook).
2. All students in the course are auto-added to course_users.

KastanDay commented on ai_ta_backend/data_logging.py:
Can we clean up this PR to be only Canvas functions? With data logging in here it'll conflict with the Nomic data logging PR.

One feature per PR, so we can merge them quickly.

star-nox replied:
Done! The Canvas functions are still a work in progess:
1. Need to do a final check for ingest and update functions
2. Noticed that the current Canvas export API only exports files and not the textual content that might be present (syllabus, course and module info), so need to work on that.

KastanDay replied:
> Done! The Canvas functions are still a work in progess:
> 
> 1. Need to do a final check for ingest and update functions
> 2. Noticed that the current Canvas export API only exports files and not the textual content that might be present (syllabus, course and module info), so need to work on that.

Sounds good, thanks for the update!

KastanDay commented:
We also want to grab all "page content" which is defined as Syllabus page, Rubrics page, maybe others too. Only if they exist. 
We want these pages if they're helpful. 

Remember the "black hole" ingest philosophy. 

<img width="1044" alt="CleanShot 2023-08-21 at 14 20 00" src="https://github.com/UIUC-Chatbot/ai-ta-backend/assets/13607221/c3f806af-9ce6-44cf-9744-2d7495a2e953">

KastanDay commented:
Adding a link to the current Canvas apps on Illinois [Canvas@Illinois, General LTI Integrations (uillinois.edu)](https://answers.uillinois.edu/illinois/87867)

KastanDay commented on ai_ta_backend/update_materials.py:
Wonderful work here, thanks @star-nox !

star-nox commented:
1. Added uiuc.chat as a Teacher to the dev course and replaced access token. Student doesn't have access to user IDs, so can't implement `add_users()` with a Student profile.
2. Need to pass the collected user emails to the function which adds users to uiuc.chat.
3. Ingest and Update functions for Canvas files  are done. Looking into scraping the page content next.

star-nox commented:
We are now able to scrape all of the content in Canvas (syllabus, pages, etc.)!

KastanDay commented:
Wow fantastic, can we do a live demo on next meeting?

KastanDay commented:
Make sure to finish any details necessary so professors can start using it. Just LMK if you have any questions, the more questions the better.

star-nox commented:
Current progress on the feature: Currently it exports all course content irrespective of whether anything is unpublished. I am working on creating a list of published content first and then passing it to the content export API, so only those files are exported.

KastanDay commented:
Shouldn’t it be fine? If the “student user” can see it, then it’s meant to
be seen?

On Mon, Sep 18, 2023 at 3:15 PM Asmita Dabholkar ***@***.***>
wrote:

> Current progress on the feature: Currently it exports all course content
> irrespective of whether anything is unpublished. I am working on creating a
> list of published content first and then passing it to the content export
> API, so only those files are exported.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UIUC-Chatbot/ai-ta-backend/pull/33#issuecomment-1724533589>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADH2CNOVIDJ7FXD4FT4SC4DX3DBZDANCNFSM6AAAAAA3MDQF5A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

star-nox commented:
> Shouldn’t it be fine? If the “student user” can see it, then it’s meant to be seen?
> […](#)
> On Mon, Sep 18, 2023 at 3:15 PM Asmita Dabholkar ***@***.***> wrote: Current progress on the feature: Currently it exports all course content irrespective of whether anything is unpublished. I am working on creating a list of published content first and then passing it to the content export API, so only those files are exported. — Reply to this email directly, view it on GitHub <[#33 (comment)](https://github.com/UIUC-Chatbot/ai-ta-backend/pull/33#issuecomment-1724533589)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2CNOVIDJ7FXD4FT4SC4DX3DBZDANCNFSM6AAAAAA3MDQF5A> . You are receiving this because you commented.Message ID: ***@***.***>

I get unauthorized user warning when i try exporting through a student profile
<img width="608" alt="image" src="https://github.com/UIUC-Chatbot/ai-ta-backend/assets/25609228/fd1a28a9-254a-4b31-8d35-f558dc59fcfe">
This works fine with a teacher profile.

KastanDay commented:
Yeah but we *need* to use the student profile, not teacher. We verified that this should work before... please try to fix.

KastanDay commented:
For the student user, I think you still need to create a `bearer auth token` in the settings, then use that token in the API requests.

KastanDay commented:
Check if you can pull data from each of these sections individually: 
![CleanShot 2023-09-19 at 14 04 18](https://github.com/UIUC-Chatbot/ai-ta-backend/assets/13607221/12d5d54a-a449-4391-a49e-4a0eb41b7739)
Docs:
![CleanShot 2023-09-19 at 14 05 04](https://github.com/UIUC-Chatbot/ai-ta-backend/assets/13607221/82efe2fe-aae7-4015-b538-9af50e8acc62)

Focus on the most relevant tabs: Pages, Files, Syllabus... etc.

star-nox commented:
> For the student user, I think you still need to create a `bearer auth token` in the Canvas GUI settings, then use that token in the API requests.

Yup, I have that created but it still says unauthorized user. 
<img width="710" alt="image" src="https://github.com/UIUC-Chatbot/ai-ta-backend/assets/25609228/6c4621ec-0a60-4e6b-b556-d9689d1fe5c0">

I can access the individual files though like you suggested, so I am trying that route. I will have to refactor the ingest function, I am working on that. Thanks for your help!

rohan-uiuc commented:
These might be helpful @star-nox 
https://community.canvaslms.com/t5/Canvas-Question-Forum/Use-API-to-export-only-pages/m-p/460735
https://community.canvaslms.com/t5/Canvas-Developers-Group/Using-Content-Migrations-API-to-copy-courses-w-o-Announcements/m-p/470539

star-nox commented:
> These might be helpful @star-nox https://community.canvaslms.com/t5/Canvas-Question-Forum/Use-API-to-export-only-pages/m-p/460735 https://community.canvaslms.com/t5/Canvas-Developers-Group/Using-Content-Migrations-API-to-copy-courses-w-o-Announcements/m-p/470539

Selective content export yes! I was trying exactly that before Kastan suggested the individual files route. I have uiuc.chat as a user in the course. I have the bearer token set. But the `content_exports` API call returns an authorization error as a student, but works as a teacher.

star-nox commented:
> Yeah but we _need_ to use the student profile, not teacher. We verified that this should work before... please try to fix.
> 
> Don't you remember how we debated endlessly between:
> 
> 1. ❌ Forcing course creators to supply their Canvas API key and
> 2. ✅ Using a student account that professors invite into their course? We need to use the student account.

Oh, by teacher I don't mean the course creators. I am referring to using the UIUC Chat account which professors will add to their course. Instead of adding it as a student, I was adding it as a teacher.

star-nox commented:
@KastanDay I have updated the ingest and add_users functions here.
I have modified the workflow of ingest such that it takes a dictionary containing boolean values for what content to ingest and acts accordingly. I am currently working on modifying the update function to reflect this workflow.

KastanDay commented:
Hi @star-nox, do you think any of this is ready for testing & review? I'd love to get an MVP live ASAP so I can give it to George to better understand what users will want. Thanks.

KastanDay commented on ai_ta_backend/vector_database.py:
```diff
@@ -298,6 +298,8 @@ def _ingest_single_py(self, s3_path: str, course_name: str):
             'base_url': '',
         } for doc in documents]
+      #print(texts)
+      os.remove(file_path)
```
Should this `os.remove()` be here twice? Seems like it could cause bugs.

star-nox replied:
Fixed this!

KastanDay commented on ai_ta_backend/update_materials.py:
```diff
@@ -0,0 +1,96 @@
+            md5_hash.update(byte_block)
+    return md5_hash.hexdigest()
+
+def update_files(source_path: str, course_name: str):
```
Oh I'm so excited to have this function!

KastanDay replied:
But please update it to also delete the old files if they're being replaced. We have a function for this (it's called in the /delete endpoint).

star-nox replied:
Okay! Right now I delete the newly downloaded file if its checksum matches with any of the existing file and upload the rest (with non-matching checksums). 
I will add an extra condition - if the checksum does not match, but the filename already exists, replace the old file with the new.

star-nox replied:
This is done!

KastanDay reviewed (changes requested):
Minor change to `update_files()`

star-nox commented:
> Hi @star-nox, do you think any of this is ready for testing & review? I'd love to get an MVP live ASAP so I can give it to George to better understand what users will want. Thanks.

I was facing some JSON error when I passed this dictionary as a parameter in the API call. It is supposed to come from the front-end. For now I have hardcoded it into the function - it ingests all of the below things. It works well for now though, you can test it.
```
            # a dictionary of all contents we want to ingest - files, pages, modules, syllabus, assignments, discussions.
            content_to_ingest = {
                'files': True,
                'pages': True,
                'modules': True,
                'syllabus': True,
                'assignments': True,
                'discussions': True
            }
```

star-nox commented:
@KastanDay `update_files()` is also ready for review.

KastanDay commented on ai_ta_backend/update_materials.py:
```diff
@@ -0,0 +1,91 @@
+    
+
+    # Access checksum of s3 files
+    s3_client = boto3.client('s3', aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
```
Hmm, this is really expensive to download all files from S3. 

In the end, it's the exact same as using the `Document Contexts` from Supabase SQL. I think we have to re-implement this with Supabase instead of S3. It'll be hundreds of times faster and cheaper and probably conceptually simpler too.

KastanDay commented on ai_ta_backend/main.py:
```diff
@@ -373,6 +375,50 @@ def mit_download_course() -> Response:
+  """
+  canvas = CanvasAPI()
+  canvas_course_id: str = request.args.get('course_id')
+  course_name: str = request.args.get('course_name')
```
Please add parameter checking to these endpoints with the `abort()`. As in `/mit-download`. Thanks.

star-nox replied:
done!

KastanDay commented on ai_ta_backend/main.py:
```diff
@@ -2,6 +2,7 @@
 import os
 import time
 from typing import List
+import json
```
Let's remove this import, it appears to be unused.

star-nox replied:
done!

KastanDay reviewed (changes requested):
Changes requested in main & update_materials.

star-nox commented:
@KastanDay I have separated the file update functions into a different PR. This one now contains the `addCanvasUsers` and `ingestCanvas` APIs which are ready for testing and demo. I have added these APIs to our Postman workspace too.

KastanDay reviewed (approved):
Fully working in my end-to-end testing. Merging!
